convert.py

```diff
@@ -297,127 +297,144 @@ def load(model_plus: ModelPlus) -> Params:
-        if expected_ids != actual_ids:
-            expected_end_id = vocab_size + len(actual_ids) - 1
-            raise Exception(f"Expected the {len(actual_ids)} added token ID(s) to be sequential in the range {vocab_size} - {expected_end_id}; got {actual_ids}")
-
-        items = sorted(added_tokens.items(), key=lambda text_idx: text_idx[1])
-        self.added_tokens_list    = [text for (text, idx) in items]
-        self.vocab_size_base: int = vocab_size
-        self.vocab_size: int      = self.vocab_size_base + len(self.added_tokens_list)
-        self.fname_tokenizer      = fname_tokenizer
-        self.fname_added_tokens   = fname_added_tokens
-
-    def bpe_tokens(self) -> Iterable[tuple[bytes, float, gguf.TokenType]]:
-        tokenizer = self.bpe_tokenizer
-        from transformers.models.gpt2 import tokenization_gpt2  # type: ignore[import]
-        reverse_vocab = {id: encoded_tok for encoded_tok, id in tokenizer.items()}
-
-        for i, _ in enumerate(tokenizer):
-            yield reverse_vocab[i], 0.0, gguf.TokenType.NORMAL
-
-    def added_tokens(self) -> Iterable[tuple[bytes, float, gguf.TokenType]]:
-        for text in self.added_tokens_list:
-            score = -1000.0
-            yield text.encode("utf-8"), score, gguf.TokenType.CONTROL
-
-    def all_tokens(self) -> Iterable[tuple[bytes, float, gguf.TokenType]]:
-        yield from self.bpe_tokens()
-        yield from self.added_tokens()
-
-    def __repr__(self) -> str:
-        return f"<BpeVocab with {self.vocab_size_base} base tokens and {len(self.added_tokens_list)} added tokens>"
-
-
-class SentencePieceVocab:
-    def __init__(self, fname_tokenizer: Path, fname_added_tokens: Path | None) -> None:
-        self.sentencepiece_tokenizer = SentencePieceProcessor(str(fname_tokenizer))
-        added_tokens: dict[str, int]
-        if fname_added_tokens is not None:
-            added_tokens = json.load(open(fname_added_tokens, encoding="utf-8"))
+class VocabLoader:
+    def __init__(self, params: Params, fname_tokenizer: Path) -> None:
+        try:
+            from transformers import AutoTokenizer
+        except ImportError as e:
+            raise ImportError(
+                "To use VocabLoader, please install the `transformers` package. "
+                "You can install it with `pip install transformers`."
+            ) from e
+
+        try:
+            self.tokenizer = AutoTokenizer.from_pretrained(str(fname_tokenizer), trust_remote_code=True)
+            vocab_set = {encoded_tok for encoded_tok, id in self.tokenizer.get_vocab().items()}
+        except:
+            self.tokenizer = AutoTokenizer.from_pretrained(str(fname_tokenizer), use_fast=False, trust_remote_code=True)
+            vocab_set = {encoded_tok for encoded_tok, id in self.tokenizer.get_vocab().items()}
+        
+        self.added_tokens_list = []
+        self.added_tokens_dict = dict()
+        self.added_tokens_ids = set()
```

Comment:
mypy won't be able to infer the types here, you'll need to annotate them:
```suggestion
        self.added_tokens_list: list[str] = []
        self.added_tokens_dict: dict[str, int] = {}
        self.added_tokens_ids: set[int] = set()
```
Though it would be more Pythonic if you created the set from the dict after the loop, e.g.
```python
self.added_tokens_ids = set(self.added_tokens_dict.values())
```
And I haven't looked closely at the code that uses these variables, but couldn't you use an OrderedDict to remove the need to store the list separately (assuming the order is important and not the ability to have duplicates)?

Reply:
@cebtenzzre 
I know that OrderedDict is supported as the default type of dict since python 3.6. Do you happen to know what version of python does llama.cpp support as the lowest version? 

If it's 3.6 or higher, I think I can handle it with dict, but if it's supported below that, I'll have to implement it similarly to now or with OrderedDict as you suggested.

Reply:
@cebtenzzre 
I thought about what you suggested and realized that it would be better to implement it as an OrderedDict, so I changed the implementation. Thanks for the suggestion.